File: old/algo/206.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# 반복
# class Solution:
#     def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         rev = None
#         while head:
#             rev, rev.next, head = head, rev, head.next
#         return rev

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def reverse(node: Optional[ListNode], prev: Optional[ListNode] = None):
            if not node:
                return prev
            prev, prev.next, node = node, prev, node.next
            return reverse(node, prev)

        return reverse(head, None)
    
list = [1,2,3,4,5]
head = ListNode(list[0])
for i in list[1:]:
    p = head
    while p.next:
        p = p.next
    p.next = ListNode(i)

rev_head = Solution().reverseList(head)
while rev_head:
    print(rev_head.val)
    rev_head = rev_head.next

File: old/algo/206.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var node = head
        var rev: ListNode? = null

        while (node != null) {
            val next = node.next
            val prev = rev
            rev = node
            rev.next = prev
            node = next
        }
        return rev
    }
}

var head = ListNode(1)
head.next = ListNode(2)
head.next?.next = ListNode(3)
head.next?.next?.next = ListNode(4)

var rev = Solution().reverseList(head)

while (rev != null) {
    println(rev?.`val`)
    rev = rev?.next
}
